### PR TITLE
Fixes #297 Modify script to create buckets based on bucket shadowing …

### LINKED ISSUE #297: Modify script to create buckets based on bucket shadowing config
Also, add trivial functional test that simply sets up a bucket shadowing config.

### PR DESCRIPTION
**Pull request by tleyden**
…config

Changes:

- Refactor out `discover_buckets_from_config`
- Refactor out `discover_mode`
- Update `discover_buckets_from_config` to consider shadow buckets

### CODE REVIEW

File: lib/cluster.py [hunk 1]
```diff
@@ -300,3 +272,81 @@ def __repr__(self):
...
+        if "cluster_config" in conf_obj.keys():
+            return  "distributed_index"
+
+    return "channel_cache"
+    
+
+def strip_invalid_json_from_config(data):
+
+    # strip out templated variables {{ ... }}
+        data = re.sub("({{.*}})", "0", data)
```
Comment (sethrosetter) [inline lib/cluster.py]: @tleyden Indentation looks a little strange here

Comment (tleyden) [inline lib/cluster.py]: oh, it's probably a tabs / spaces thing.  I'll fix it, thanks for catching.

File: lib/cluster.py [hunk 2]
```diff
@@ -300,3 +272,81 @@ def __repr__(self):
...
+    # Buckets may be shared for different functionality
+    bucket_name_set = list(set(bucket_names_from_config))
+
+    return bucket_name_set
+
+    
+if __name__ == "__main__":
+
+    # poor man's unit test
+    conf_path = "conf/sync_gateway_bucketshadow_cc.json"
```
Comment (sethrosetter) [inline lib/cluster.py]: I think we should do this a bit different. Maybe we could return mode and buckets from the reset method. We may need to to that in the future anyway for config testing.

Comment (tleyden) [inline lib/cluster.py]: What about something like this?

```
config_wrapper = ConfigWrapper(conf_path)
for bucket in config_wrapper.buckets:
    ...
if config_wrapper.mode == DIST_INDEX:
    ...
if len(config_wrapper.num_databases) >1:
    ...

```

File: lib/cluster.py (hunk 1 above)
Comment (tleyden) [inline lib/cluster.py]: wasn't a tab / spaces thing, just needed some more indents.  committed a fix locally.

**Comment by tleyden** [comment]
@sethrosetter I re-pushed with the changes we discussed -- take a look.

```
py.test -s "functional_tests/test_bucket_shadow.py"
```

is passing, but I have not tried to retest the other tests

**Comment by tleyden** [comment]
Updated PR to remove the `test_bucket_shadow.py` file, since that is still in progress

**Comment by sethrosetter** [comment]
Test run succeeded.

**Pull request merged**